lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,232 @@
+            runtime.extend(&[NodeHasLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        }
+    }
+
+    let (accesses, deny_accesses, deny_logins) = runtime.run();
+    let mut predicates = vec![];
+    predicates.extend(
+        accesses
+            .into_iter()
+            .map(|HasAccess(a, b, c, d)| types::facts::Predicate {
+                name: types::facts::PredicateType::HasAccess as i32,
+                atoms: vec![a, b, c, d],
+            }),
+    );
+    predicates.extend(
+        deny_accesses
+            .into_iter()
+            .map(|DenyAccess(a, b, c, d)| types::facts::Predicate {
+                name: types::facts::PredicateType::DenyAccess as i32,
+                atoms: vec![a, b, c, d],
+            })
+    );
+    predicates.extend(
+        deny_logins
+            .into_iter()
+            .map(|DenyLogins(a, b, c)| types::facts::Predicate {
+                name: types::facts::PredicateType::DenyLogins as i32,
+                atoms: vec![a, b, c],
+            })
+    );
+
+    let idb = types::Facts {
+        predicates,
+    };
+
+    let mut buf = Vec::with_capacity(idb.encoded_len());
+    match idb.encode(&mut buf) {
+        Ok(_) => (),
+        Err(e) => {
+            let err_bytes = e.to_string().into_bytes();
+            unsafe {
+                ptr::copy(&(*err_bytes)[0], output, err_bytes.len());
+            }
+            return Box::into_raw(Box::new(Status{
+                output_length: err_bytes.len(),
+                error: -1,
+            }))
+        },
+    };
+    if buf.is_empty() || buf.len() > output_len {
+        return Box::into_raw(Box::new(Status{
+            output_length: buf.len(),
+            error: 0,
+        }))
+    }
+    // We can't guarantee the memory regions are non-overlapping, but we only need to copy the data to output so
+    // it is able to be presented on the Go end. The necessary checks for length is done before this call.
+    //
+    // buf is valid for reads of len * size_of::<T>() bytes.
+    // output is valid for writes of count * size_of::<T>() bytes.
+    // Both buf and output are properly aligned.
+    unsafe {
+        ptr::copy(&(*buf)[0], output, buf.len());
+    }
+
+    Box::into_raw(Box::new(Status{
+        output_length: buf.len(),
+        error: 0
+    }))
+}
+
+#[no_mangle]
+pub extern "C" fn status_output_length(
+    status: *mut Status
+) -> size_t {
+    if status.is_null() {
+        return 0
+    }
+    // We've checked that the pointer is not null.
+    let db = unsafe {
+        &*status
+    };
+    db.output_length
+}
+
+#[no_mangle]
+pub extern "C" fn status_error(
+    status: *mut Status
+) -> i32 {
+    if status.is_null() {
+        return 0
+    }
+    // We've checked that the pointer is not null.
+    let db = unsafe {
+        &*status
+    };
+    db.error
+}
+
+#[no_mangle]
+extern "C" fn drop_rust_buffer(buf: *mut c_uchar, len: size_t) {
```

Comment:
Is this intended to drop a `Vec<T>`? If so, it doesn't, you're merely creating a slice which is a view into memory here, it doesn't have a drop implementation since it doesn't own any resources. If you're wanting to deallocate a `Vec<c_uchar>` you'll need it's length, capacity and base pointer and create the vec using it's `from_parts` method and assign it to `let _ = expr`. No need for `mem::drop`.